site/app/models/gradeable/Gradeable.php

```diff
@@ -0,0 +1,446 @@
+                $errors[$date] = $result;
+            }
+        }
+
+        $ta_view_start_date = $dates['ta_view_start_date'];
+        $grade_start_date = $dates['grade_start_date'];
+        $grade_released_date = $dates['grade_released_date'];
+        $team_lock_date = $dates['team_lock_date'];
+        $submission_open_date = $dates['submission_open_date'];
+        $submission_due_date = $dates['submission_due_date'];
+        $late_days = $dates['late_days'];
+
+        $late_interval = null;
+        $late_days = intval($late_days);
+        if ($late_days < 0) {
+            $errors['late_days'] = 'Late day count must be >= 0!';
+        } else {
+            try {
+                $late_interval = new \DateInterval('P' . strval($late_days) . 'D');
+            }
+            catch (\Exception $e) {
+                // This is for development debugging. In reality, we should never hit this line
+                $errors['late_days'] = "Error parsing late days: {$e}";
+            }
+        }
+
+        $max_due = $submission_due_date;
+        if ($submission_due_date instanceof \DateTime && $late_interval !== null) {
+            $max_due = (clone $submission_due_date)->add($late_interval);
+        }
+
+        if ($ta_view_start_date === null) {
+            $errors['ta_view_start_date'] = "Value must not be null!";
+        }
+        if ($grade_released_date === null) {
+            $errors['grade_released_date'] = "Value must not be null!";
+        }
+
+        if ($this->type === GradeableType::ELECTRONIC_FILE) {
+            if ($submission_open_date === null) {
+                $errors['submission_open_date'] = "Value must not be null!";
+            }
+            if ($submission_due_date === null) {
+                $errors['submission_due_date'] = "Value must not be null!";
+            }
+
+            if (!($ta_view_start_date === null || $submission_open_date === null) && $ta_view_start_date > $submission_open_date) {
+                $errors['g_ta_view_start_date'] = 'TA Beta Testing Date must not be later than Submission Open Date';
+            }
+            if (!($submission_open_date === null || $submission_due_date === null) && $submission_open_date > $submission_due_date) {
+                $errors['eg_submission_open_date'] = 'Submission Open Date must not be later than Submission Due Date';
+            }
+            if ($this->ta_grading) {
+                if ($grade_start_date === null) {
+                    $errors['grade_start_date'] = "Value must not be null!";
+                }
+//                if ($grade_locked_date === null) {
+//                    $errors['grade_locked_date'] = "Value must not be null!";
+//                }
+                if (!($submission_due_date === null || $grade_start_date === null) && $submission_due_date > $grade_start_date) {
+                    $errors['g_grade_start_date'] = 'Manual Grading Open Date must be no earlier than Due Date';
```

Comment:
Nitpick: single or double quotes for these error messages?

Reply:
Single quotes used unless double quotes needed